// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/vault/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.OperatorParams = exports.VaultParams = exports.QuotingParams = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseQuotingParams() {
    return {
        layers: 0,
        spreadMinPpm: 0,
        spreadBufferPpm: 0,
        skewFactorPpm: 0,
        orderSizePctPpm: 0,
        orderExpirationSeconds: 0,
        activationThresholdQuoteQuantums: new Uint8Array()
    };
}
exports.QuotingParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.layers !== 0) {
            writer.uint32(8).uint32(message.layers);
        }
        if (message.spreadMinPpm !== 0) {
            writer.uint32(16).uint32(message.spreadMinPpm);
        }
        if (message.spreadBufferPpm !== 0) {
            writer.uint32(24).uint32(message.spreadBufferPpm);
        }
        if (message.skewFactorPpm !== 0) {
            writer.uint32(32).uint32(message.skewFactorPpm);
        }
        if (message.orderSizePctPpm !== 0) {
            writer.uint32(40).uint32(message.orderSizePctPpm);
        }
        if (message.orderExpirationSeconds !== 0) {
            writer.uint32(48).uint32(message.orderExpirationSeconds);
        }
        if (message.activationThresholdQuoteQuantums.length !== 0) {
            writer.uint32(58).bytes(message.activationThresholdQuoteQuantums);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQuotingParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.layers = reader.uint32();
                    break;
                case 2:
                    message.spreadMinPpm = reader.uint32();
                    break;
                case 3:
                    message.spreadBufferPpm = reader.uint32();
                    break;
                case 4:
                    message.skewFactorPpm = reader.uint32();
                    break;
                case 5:
                    message.orderSizePctPpm = reader.uint32();
                    break;
                case 6:
                    message.orderExpirationSeconds = reader.uint32();
                    break;
                case 7:
                    message.activationThresholdQuoteQuantums = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseQuotingParams();
        message.layers = (_a = object.layers) !== null && _a !== void 0 ? _a : 0;
        message.spreadMinPpm = (_b = object.spreadMinPpm) !== null && _b !== void 0 ? _b : 0;
        message.spreadBufferPpm = (_c = object.spreadBufferPpm) !== null && _c !== void 0 ? _c : 0;
        message.skewFactorPpm = (_d = object.skewFactorPpm) !== null && _d !== void 0 ? _d : 0;
        message.orderSizePctPpm = (_e = object.orderSizePctPpm) !== null && _e !== void 0 ? _e : 0;
        message.orderExpirationSeconds = (_f = object.orderExpirationSeconds) !== null && _f !== void 0 ? _f : 0;
        message.activationThresholdQuoteQuantums = (_g = object.activationThresholdQuoteQuantums) !== null && _g !== void 0 ? _g : new Uint8Array();
        return message;
    }
};
function createBaseVaultParams() {
    return {
        status: 0,
        quotingParams: undefined
    };
}
exports.VaultParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.status !== 0) {
            writer.uint32(8).int32(message.status);
        }
        if (message.quotingParams !== undefined) {
            exports.QuotingParams.encode(message.quotingParams, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseVaultParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.status = reader.int32();
                    break;
                case 2:
                    message.quotingParams = exports.QuotingParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseVaultParams();
        message.status = (_a = object.status) !== null && _a !== void 0 ? _a : 0;
        message.quotingParams = object.quotingParams !== undefined && object.quotingParams !== null ? exports.QuotingParams.fromPartial(object.quotingParams) : undefined;
        return message;
    }
};
function createBaseOperatorParams() {
    return {
        operator: ""
    };
}
exports.OperatorParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.operator !== "") {
            writer.uint32(10).string(message.operator);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOperatorParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.operator = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseOperatorParams();
        message.operator = (_a = object.operator) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
function createBaseParams() {
    return {
        layers: 0,
        spreadMinPpm: 0,
        spreadBufferPpm: 0,
        skewFactorPpm: 0,
        orderSizePctPpm: 0,
        orderExpirationSeconds: 0,
        activationThresholdQuoteQuantums: new Uint8Array()
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.layers !== 0) {
            writer.uint32(8).uint32(message.layers);
        }
        if (message.spreadMinPpm !== 0) {
            writer.uint32(16).uint32(message.spreadMinPpm);
        }
        if (message.spreadBufferPpm !== 0) {
            writer.uint32(24).uint32(message.spreadBufferPpm);
        }
        if (message.skewFactorPpm !== 0) {
            writer.uint32(32).uint32(message.skewFactorPpm);
        }
        if (message.orderSizePctPpm !== 0) {
            writer.uint32(40).uint32(message.orderSizePctPpm);
        }
        if (message.orderExpirationSeconds !== 0) {
            writer.uint32(48).uint32(message.orderExpirationSeconds);
        }
        if (message.activationThresholdQuoteQuantums.length !== 0) {
            writer.uint32(58).bytes(message.activationThresholdQuoteQuantums);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.layers = reader.uint32();
                    break;
                case 2:
                    message.spreadMinPpm = reader.uint32();
                    break;
                case 3:
                    message.spreadBufferPpm = reader.uint32();
                    break;
                case 4:
                    message.skewFactorPpm = reader.uint32();
                    break;
                case 5:
                    message.orderSizePctPpm = reader.uint32();
                    break;
                case 6:
                    message.orderExpirationSeconds = reader.uint32();
                    break;
                case 7:
                    message.activationThresholdQuoteQuantums = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseParams();
        message.layers = (_a = object.layers) !== null && _a !== void 0 ? _a : 0;
        message.spreadMinPpm = (_b = object.spreadMinPpm) !== null && _b !== void 0 ? _b : 0;
        message.spreadBufferPpm = (_c = object.spreadBufferPpm) !== null && _c !== void 0 ? _c : 0;
        message.skewFactorPpm = (_d = object.skewFactorPpm) !== null && _d !== void 0 ? _d : 0;
        message.orderSizePctPpm = (_e = object.orderSizePctPpm) !== null && _e !== void 0 ? _e : 0;
        message.orderExpirationSeconds = (_f = object.orderExpirationSeconds) !== null && _f !== void 0 ? _f : 0;
        message.activationThresholdQuoteQuantums = (_g = object.activationThresholdQuoteQuantums) !== null && _g !== void 0 ? _g : new Uint8Array();
        return message;
    }
};
