// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/vault/query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryMegavaultWithdrawalInfoResponse = exports.QueryMegavaultWithdrawalInfoRequest = exports.QueryVaultParamsResponse = exports.QueryVaultParamsRequest = exports.QueryMegavaultAllOwnerSharesResponse = exports.QueryMegavaultAllOwnerSharesRequest = exports.QueryMegavaultOwnerSharesResponse = exports.QueryMegavaultOwnerSharesRequest = exports.QueryMegavaultTotalSharesResponse = exports.QueryMegavaultTotalSharesRequest = exports.QueryAllVaultsResponse = exports.QueryAllVaultsRequest = exports.QueryVaultResponse = exports.QueryVaultRequest = exports.QueryParamsResponse = exports.QueryParamsRequest = void 0;
const vault_1 = require("./vault");
const pagination_1 = require("../../cosmos/base/query/v1beta1/pagination");
const share_1 = require("./share");
const params_1 = require("./params");
const subaccount_1 = require("../subaccounts/subaccount");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseQueryParamsRequest() {
    return {};
}
exports.QueryParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryParamsRequest();
        return message;
    }
};
function createBaseQueryParamsResponse() {
    return {
        params: undefined,
        defaultQuotingParams: undefined,
        operatorParams: undefined
    };
}
exports.QueryParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.params !== undefined) {
            params_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        if (message.defaultQuotingParams !== undefined) {
            params_1.QuotingParams.encode(message.defaultQuotingParams, writer.uint32(18).fork()).ldelim();
        }
        if (message.operatorParams !== undefined) {
            params_1.OperatorParams.encode(message.operatorParams, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = params_1.Params.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.defaultQuotingParams = params_1.QuotingParams.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.operatorParams = params_1.OperatorParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? params_1.Params.fromPartial(object.params) : undefined;
        message.defaultQuotingParams = object.defaultQuotingParams !== undefined && object.defaultQuotingParams !== null ? params_1.QuotingParams.fromPartial(object.defaultQuotingParams) : undefined;
        message.operatorParams = object.operatorParams !== undefined && object.operatorParams !== null ? params_1.OperatorParams.fromPartial(object.operatorParams) : undefined;
        return message;
    }
};
function createBaseQueryVaultRequest() {
    return {
        type: 0,
        number: 0
    };
}
exports.QueryVaultRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.type !== 0) {
            writer.uint32(8).int32(message.type);
        }
        if (message.number !== 0) {
            writer.uint32(16).uint32(message.number);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryVaultRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.type = reader.int32();
                    break;
                case 2:
                    message.number = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseQueryVaultRequest();
        message.type = (_a = object.type) !== null && _a !== void 0 ? _a : 0;
        message.number = (_b = object.number) !== null && _b !== void 0 ? _b : 0;
        return message;
    }
};
function createBaseQueryVaultResponse() {
    return {
        vaultId: undefined,
        subaccountId: undefined,
        equity: new Uint8Array(),
        inventory: new Uint8Array(),
        vaultParams: undefined,
        mostRecentClientIds: []
    };
}
exports.QueryVaultResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.vaultId !== undefined) {
            vault_1.VaultId.encode(message.vaultId, writer.uint32(10).fork()).ldelim();
        }
        if (message.subaccountId !== undefined) {
            subaccount_1.SubaccountId.encode(message.subaccountId, writer.uint32(18).fork()).ldelim();
        }
        if (message.equity.length !== 0) {
            writer.uint32(26).bytes(message.equity);
        }
        if (message.inventory.length !== 0) {
            writer.uint32(34).bytes(message.inventory);
        }
        if (message.vaultParams !== undefined) {
            params_1.VaultParams.encode(message.vaultParams, writer.uint32(42).fork()).ldelim();
        }
        writer.uint32(50).fork();
        for (const v of message.mostRecentClientIds) {
            writer.uint32(v);
        }
        writer.ldelim();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryVaultResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.vaultId = vault_1.VaultId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.subaccountId = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.equity = reader.bytes();
                    break;
                case 4:
                    message.inventory = reader.bytes();
                    break;
                case 5:
                    message.vaultParams = params_1.VaultParams.decode(reader, reader.uint32());
                    break;
                case 6:
                    if ((tag & 7) === 2) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.mostRecentClientIds.push(reader.uint32());
                        }
                    }
                    else {
                        message.mostRecentClientIds.push(reader.uint32());
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseQueryVaultResponse();
        message.vaultId = object.vaultId !== undefined && object.vaultId !== null ? vault_1.VaultId.fromPartial(object.vaultId) : undefined;
        message.subaccountId = object.subaccountId !== undefined && object.subaccountId !== null ? subaccount_1.SubaccountId.fromPartial(object.subaccountId) : undefined;
        message.equity = (_a = object.equity) !== null && _a !== void 0 ? _a : new Uint8Array();
        message.inventory = (_b = object.inventory) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.vaultParams = object.vaultParams !== undefined && object.vaultParams !== null ? params_1.VaultParams.fromPartial(object.vaultParams) : undefined;
        message.mostRecentClientIds = ((_c = object.mostRecentClientIds) === null || _c === void 0 ? void 0 : _c.map(e => e)) || [];
        return message;
    }
};
function createBaseQueryAllVaultsRequest() {
    return {
        pagination: undefined
    };
}
exports.QueryAllVaultsRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAllVaultsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryAllVaultsRequest();
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageRequest.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseQueryAllVaultsResponse() {
    return {
        vaults: [],
        pagination: undefined
    };
}
exports.QueryAllVaultsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.vaults) {
            exports.QueryVaultResponse.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAllVaultsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.vaults.push(exports.QueryVaultResponse.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryAllVaultsResponse();
        message.vaults = ((_a = object.vaults) === null || _a === void 0 ? void 0 : _a.map(e => exports.QueryVaultResponse.fromPartial(e))) || [];
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageResponse.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseQueryMegavaultTotalSharesRequest() {
    return {};
}
exports.QueryMegavaultTotalSharesRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultTotalSharesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryMegavaultTotalSharesRequest();
        return message;
    }
};
function createBaseQueryMegavaultTotalSharesResponse() {
    return {
        totalShares: undefined
    };
}
exports.QueryMegavaultTotalSharesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.totalShares !== undefined) {
            share_1.NumShares.encode(message.totalShares, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultTotalSharesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.totalShares = share_1.NumShares.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryMegavaultTotalSharesResponse();
        message.totalShares = object.totalShares !== undefined && object.totalShares !== null ? share_1.NumShares.fromPartial(object.totalShares) : undefined;
        return message;
    }
};
function createBaseQueryMegavaultOwnerSharesRequest() {
    return {
        address: ""
    };
}
exports.QueryMegavaultOwnerSharesRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultOwnerSharesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.address = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryMegavaultOwnerSharesRequest();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
function createBaseQueryMegavaultOwnerSharesResponse() {
    return {
        address: "",
        shares: undefined,
        shareUnlocks: [],
        equity: new Uint8Array(),
        withdrawableEquity: new Uint8Array()
    };
}
exports.QueryMegavaultOwnerSharesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        if (message.shares !== undefined) {
            share_1.NumShares.encode(message.shares, writer.uint32(18).fork()).ldelim();
        }
        for (const v of message.shareUnlocks) {
            share_1.ShareUnlock.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (message.equity.length !== 0) {
            writer.uint32(34).bytes(message.equity);
        }
        if (message.withdrawableEquity.length !== 0) {
            writer.uint32(42).bytes(message.withdrawableEquity);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultOwnerSharesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.address = reader.string();
                    break;
                case 2:
                    message.shares = share_1.NumShares.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.shareUnlocks.push(share_1.ShareUnlock.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.equity = reader.bytes();
                    break;
                case 5:
                    message.withdrawableEquity = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseQueryMegavaultOwnerSharesResponse();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.shares = object.shares !== undefined && object.shares !== null ? share_1.NumShares.fromPartial(object.shares) : undefined;
        message.shareUnlocks = ((_b = object.shareUnlocks) === null || _b === void 0 ? void 0 : _b.map(e => share_1.ShareUnlock.fromPartial(e))) || [];
        message.equity = (_c = object.equity) !== null && _c !== void 0 ? _c : new Uint8Array();
        message.withdrawableEquity = (_d = object.withdrawableEquity) !== null && _d !== void 0 ? _d : new Uint8Array();
        return message;
    }
};
function createBaseQueryMegavaultAllOwnerSharesRequest() {
    return {
        pagination: undefined
    };
}
exports.QueryMegavaultAllOwnerSharesRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultAllOwnerSharesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryMegavaultAllOwnerSharesRequest();
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageRequest.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseQueryMegavaultAllOwnerSharesResponse() {
    return {
        ownerShares: [],
        pagination: undefined
    };
}
exports.QueryMegavaultAllOwnerSharesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.ownerShares) {
            share_1.OwnerShare.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultAllOwnerSharesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.ownerShares.push(share_1.OwnerShare.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryMegavaultAllOwnerSharesResponse();
        message.ownerShares = ((_a = object.ownerShares) === null || _a === void 0 ? void 0 : _a.map(e => share_1.OwnerShare.fromPartial(e))) || [];
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageResponse.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseQueryVaultParamsRequest() {
    return {
        type: 0,
        number: 0
    };
}
exports.QueryVaultParamsRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.type !== 0) {
            writer.uint32(8).int32(message.type);
        }
        if (message.number !== 0) {
            writer.uint32(16).uint32(message.number);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryVaultParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.type = reader.int32();
                    break;
                case 2:
                    message.number = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseQueryVaultParamsRequest();
        message.type = (_a = object.type) !== null && _a !== void 0 ? _a : 0;
        message.number = (_b = object.number) !== null && _b !== void 0 ? _b : 0;
        return message;
    }
};
function createBaseQueryVaultParamsResponse() {
    return {
        vaultId: undefined,
        vaultParams: undefined
    };
}
exports.QueryVaultParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.vaultId !== undefined) {
            vault_1.VaultId.encode(message.vaultId, writer.uint32(10).fork()).ldelim();
        }
        if (message.vaultParams !== undefined) {
            params_1.VaultParams.encode(message.vaultParams, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryVaultParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.vaultId = vault_1.VaultId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.vaultParams = params_1.VaultParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryVaultParamsResponse();
        message.vaultId = object.vaultId !== undefined && object.vaultId !== null ? vault_1.VaultId.fromPartial(object.vaultId) : undefined;
        message.vaultParams = object.vaultParams !== undefined && object.vaultParams !== null ? params_1.VaultParams.fromPartial(object.vaultParams) : undefined;
        return message;
    }
};
function createBaseQueryMegavaultWithdrawalInfoRequest() {
    return {
        sharesToWithdraw: undefined
    };
}
exports.QueryMegavaultWithdrawalInfoRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sharesToWithdraw !== undefined) {
            share_1.NumShares.encode(message.sharesToWithdraw, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultWithdrawalInfoRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sharesToWithdraw = share_1.NumShares.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryMegavaultWithdrawalInfoRequest();
        message.sharesToWithdraw = object.sharesToWithdraw !== undefined && object.sharesToWithdraw !== null ? share_1.NumShares.fromPartial(object.sharesToWithdraw) : undefined;
        return message;
    }
};
function createBaseQueryMegavaultWithdrawalInfoResponse() {
    return {
        sharesToWithdraw: undefined,
        expectedQuoteQuantums: new Uint8Array(),
        megavaultEquity: new Uint8Array(),
        totalShares: undefined
    };
}
exports.QueryMegavaultWithdrawalInfoResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sharesToWithdraw !== undefined) {
            share_1.NumShares.encode(message.sharesToWithdraw, writer.uint32(10).fork()).ldelim();
        }
        if (message.expectedQuoteQuantums.length !== 0) {
            writer.uint32(18).bytes(message.expectedQuoteQuantums);
        }
        if (message.megavaultEquity.length !== 0) {
            writer.uint32(26).bytes(message.megavaultEquity);
        }
        if (message.totalShares !== undefined) {
            share_1.NumShares.encode(message.totalShares, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMegavaultWithdrawalInfoResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sharesToWithdraw = share_1.NumShares.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.expectedQuoteQuantums = reader.bytes();
                    break;
                case 3:
                    message.megavaultEquity = reader.bytes();
                    break;
                case 4:
                    message.totalShares = share_1.NumShares.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseQueryMegavaultWithdrawalInfoResponse();
        message.sharesToWithdraw = object.sharesToWithdraw !== undefined && object.sharesToWithdraw !== null ? share_1.NumShares.fromPartial(object.sharesToWithdraw) : undefined;
        message.expectedQuoteQuantums = (_a = object.expectedQuoteQuantums) !== null && _a !== void 0 ? _a : new Uint8Array();
        message.megavaultEquity = (_b = object.megavaultEquity) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.totalShares = object.totalShares !== undefined && object.totalShares !== null ? share_1.NumShares.fromPartial(object.totalShares) : undefined;
        return message;
    }
};
