// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/clob/query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamTakerOrderStatus = exports.StreamTakerOrder = exports.StreamOrderbookFill = exports.StreamOrderbookUpdate = exports.StreamUpdate = exports.StreamOrderbookUpdatesResponse = exports.StreamOrderbookUpdatesRequest = exports.QueryLiquidationsConfigurationResponse = exports.QueryLiquidationsConfigurationRequest = exports.QueryStatefulOrderResponse = exports.QueryStatefulOrderRequest = exports.QueryBlockRateLimitConfigurationResponse = exports.QueryBlockRateLimitConfigurationRequest = exports.QueryEquityTierLimitConfigurationResponse = exports.QueryEquityTierLimitConfigurationRequest = exports.MevNodeToNodeCalculationResponse_MevAndVolumePerClob = exports.MevNodeToNodeCalculationResponse = exports.MevNodeToNodeCalculationRequest = exports.QueryClobPairAllResponse = exports.QueryAllClobPairRequest = exports.QueryClobPairResponse = exports.QueryGetClobPairRequest = void 0;
const pagination_1 = require("../../cosmos/base/query/v1beta1/pagination");
const mev_1 = require("./mev");
const order_1 = require("./order");
const subaccount_1 = require("../subaccounts/subaccount");
const clob_pair_1 = require("./clob_pair");
const equity_tier_limit_config_1 = require("./equity_tier_limit_config");
const block_rate_limit_config_1 = require("./block_rate_limit_config");
const liquidations_config_1 = require("./liquidations_config");
const streaming_1 = require("../subaccounts/streaming");
const off_chain_updates_1 = require("../indexer/off_chain_updates/off_chain_updates");
const matches_1 = require("./matches");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
function createBaseQueryGetClobPairRequest() {
    return {
        id: 0
    };
}
exports.QueryGetClobPairRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== 0) {
            writer.uint32(8).uint32(message.id);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryGetClobPairRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryGetClobPairRequest();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
};
function createBaseQueryClobPairResponse() {
    return {
        clobPair: undefined
    };
}
exports.QueryClobPairResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.clobPair !== undefined) {
            clob_pair_1.ClobPair.encode(message.clobPair, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryClobPairResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clobPair = clob_pair_1.ClobPair.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryClobPairResponse();
        message.clobPair = object.clobPair !== undefined && object.clobPair !== null ? clob_pair_1.ClobPair.fromPartial(object.clobPair) : undefined;
        return message;
    }
};
function createBaseQueryAllClobPairRequest() {
    return {
        pagination: undefined
    };
}
exports.QueryAllClobPairRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAllClobPairRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryAllClobPairRequest();
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageRequest.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseQueryClobPairAllResponse() {
    return {
        clobPair: [],
        pagination: undefined
    };
}
exports.QueryClobPairAllResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.clobPair) {
            clob_pair_1.ClobPair.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryClobPairAllResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clobPair.push(clob_pair_1.ClobPair.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryClobPairAllResponse();
        message.clobPair = ((_a = object.clobPair) === null || _a === void 0 ? void 0 : _a.map(e => clob_pair_1.ClobPair.fromPartial(e))) || [];
        message.pagination = object.pagination !== undefined && object.pagination !== null ? pagination_1.PageResponse.fromPartial(object.pagination) : undefined;
        return message;
    }
};
function createBaseMevNodeToNodeCalculationRequest() {
    return {
        blockProposerMatches: undefined,
        validatorMevMetrics: undefined
    };
}
exports.MevNodeToNodeCalculationRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.blockProposerMatches !== undefined) {
            mev_1.ValidatorMevMatches.encode(message.blockProposerMatches, writer.uint32(10).fork()).ldelim();
        }
        if (message.validatorMevMetrics !== undefined) {
            mev_1.MevNodeToNodeMetrics.encode(message.validatorMevMetrics, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMevNodeToNodeCalculationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.blockProposerMatches = mev_1.ValidatorMevMatches.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.validatorMevMetrics = mev_1.MevNodeToNodeMetrics.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseMevNodeToNodeCalculationRequest();
        message.blockProposerMatches = object.blockProposerMatches !== undefined && object.blockProposerMatches !== null ? mev_1.ValidatorMevMatches.fromPartial(object.blockProposerMatches) : undefined;
        message.validatorMevMetrics = object.validatorMevMetrics !== undefined && object.validatorMevMetrics !== null ? mev_1.MevNodeToNodeMetrics.fromPartial(object.validatorMevMetrics) : undefined;
        return message;
    }
};
function createBaseMevNodeToNodeCalculationResponse() {
    return {
        results: []
    };
}
exports.MevNodeToNodeCalculationResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.results) {
            exports.MevNodeToNodeCalculationResponse_MevAndVolumePerClob.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMevNodeToNodeCalculationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.results.push(exports.MevNodeToNodeCalculationResponse_MevAndVolumePerClob.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMevNodeToNodeCalculationResponse();
        message.results = ((_a = object.results) === null || _a === void 0 ? void 0 : _a.map(e => exports.MevNodeToNodeCalculationResponse_MevAndVolumePerClob.fromPartial(e))) || [];
        return message;
    }
};
function createBaseMevNodeToNodeCalculationResponse_MevAndVolumePerClob() {
    return {
        clobPairId: 0,
        mev: 0,
        volume: helpers_1.Long.UZERO
    };
}
exports.MevNodeToNodeCalculationResponse_MevAndVolumePerClob = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.clobPairId !== 0) {
            writer.uint32(8).uint32(message.clobPairId);
        }
        if (message.mev !== 0) {
            writer.uint32(21).float(message.mev);
        }
        if (!message.volume.isZero()) {
            writer.uint32(24).uint64(message.volume);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMevNodeToNodeCalculationResponse_MevAndVolumePerClob();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clobPairId = reader.uint32();
                    break;
                case 2:
                    message.mev = reader.float();
                    break;
                case 3:
                    message.volume = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMevNodeToNodeCalculationResponse_MevAndVolumePerClob();
        message.clobPairId = (_a = object.clobPairId) !== null && _a !== void 0 ? _a : 0;
        message.mev = (_b = object.mev) !== null && _b !== void 0 ? _b : 0;
        message.volume = object.volume !== undefined && object.volume !== null ? helpers_1.Long.fromValue(object.volume) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseQueryEquityTierLimitConfigurationRequest() {
    return {};
}
exports.QueryEquityTierLimitConfigurationRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryEquityTierLimitConfigurationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryEquityTierLimitConfigurationRequest();
        return message;
    }
};
function createBaseQueryEquityTierLimitConfigurationResponse() {
    return {
        equityTierLimitConfig: undefined
    };
}
exports.QueryEquityTierLimitConfigurationResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.equityTierLimitConfig !== undefined) {
            equity_tier_limit_config_1.EquityTierLimitConfiguration.encode(message.equityTierLimitConfig, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryEquityTierLimitConfigurationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.equityTierLimitConfig = equity_tier_limit_config_1.EquityTierLimitConfiguration.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryEquityTierLimitConfigurationResponse();
        message.equityTierLimitConfig = object.equityTierLimitConfig !== undefined && object.equityTierLimitConfig !== null ? equity_tier_limit_config_1.EquityTierLimitConfiguration.fromPartial(object.equityTierLimitConfig) : undefined;
        return message;
    }
};
function createBaseQueryBlockRateLimitConfigurationRequest() {
    return {};
}
exports.QueryBlockRateLimitConfigurationRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBlockRateLimitConfigurationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryBlockRateLimitConfigurationRequest();
        return message;
    }
};
function createBaseQueryBlockRateLimitConfigurationResponse() {
    return {
        blockRateLimitConfig: undefined
    };
}
exports.QueryBlockRateLimitConfigurationResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.blockRateLimitConfig !== undefined) {
            block_rate_limit_config_1.BlockRateLimitConfiguration.encode(message.blockRateLimitConfig, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBlockRateLimitConfigurationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.blockRateLimitConfig = block_rate_limit_config_1.BlockRateLimitConfiguration.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryBlockRateLimitConfigurationResponse();
        message.blockRateLimitConfig = object.blockRateLimitConfig !== undefined && object.blockRateLimitConfig !== null ? block_rate_limit_config_1.BlockRateLimitConfiguration.fromPartial(object.blockRateLimitConfig) : undefined;
        return message;
    }
};
function createBaseQueryStatefulOrderRequest() {
    return {
        orderId: undefined
    };
}
exports.QueryStatefulOrderRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.orderId !== undefined) {
            order_1.OrderId.encode(message.orderId, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryStatefulOrderRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.orderId = order_1.OrderId.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryStatefulOrderRequest();
        message.orderId = object.orderId !== undefined && object.orderId !== null ? order_1.OrderId.fromPartial(object.orderId) : undefined;
        return message;
    }
};
function createBaseQueryStatefulOrderResponse() {
    return {
        orderPlacement: undefined,
        fillAmount: helpers_1.Long.UZERO,
        triggered: false
    };
}
exports.QueryStatefulOrderResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.orderPlacement !== undefined) {
            order_1.LongTermOrderPlacement.encode(message.orderPlacement, writer.uint32(10).fork()).ldelim();
        }
        if (!message.fillAmount.isZero()) {
            writer.uint32(16).uint64(message.fillAmount);
        }
        if (message.triggered === true) {
            writer.uint32(24).bool(message.triggered);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryStatefulOrderResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.orderPlacement = order_1.LongTermOrderPlacement.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.fillAmount = reader.uint64();
                    break;
                case 3:
                    message.triggered = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryStatefulOrderResponse();
        message.orderPlacement = object.orderPlacement !== undefined && object.orderPlacement !== null ? order_1.LongTermOrderPlacement.fromPartial(object.orderPlacement) : undefined;
        message.fillAmount = object.fillAmount !== undefined && object.fillAmount !== null ? helpers_1.Long.fromValue(object.fillAmount) : helpers_1.Long.UZERO;
        message.triggered = (_a = object.triggered) !== null && _a !== void 0 ? _a : false;
        return message;
    }
};
function createBaseQueryLiquidationsConfigurationRequest() {
    return {};
}
exports.QueryLiquidationsConfigurationRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLiquidationsConfigurationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryLiquidationsConfigurationRequest();
        return message;
    }
};
function createBaseQueryLiquidationsConfigurationResponse() {
    return {
        liquidationsConfig: undefined
    };
}
exports.QueryLiquidationsConfigurationResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.liquidationsConfig !== undefined) {
            liquidations_config_1.LiquidationsConfig.encode(message.liquidationsConfig, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLiquidationsConfigurationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.liquidationsConfig = liquidations_config_1.LiquidationsConfig.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryLiquidationsConfigurationResponse();
        message.liquidationsConfig = object.liquidationsConfig !== undefined && object.liquidationsConfig !== null ? liquidations_config_1.LiquidationsConfig.fromPartial(object.liquidationsConfig) : undefined;
        return message;
    }
};
function createBaseStreamOrderbookUpdatesRequest() {
    return {
        clobPairId: [],
        subaccountIds: []
    };
}
exports.StreamOrderbookUpdatesRequest = {
    encode(message, writer = _m0.Writer.create()) {
        writer.uint32(10).fork();
        for (const v of message.clobPairId) {
            writer.uint32(v);
        }
        writer.ldelim();
        for (const v of message.subaccountIds) {
            subaccount_1.SubaccountId.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamOrderbookUpdatesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if ((tag & 7) === 2) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.clobPairId.push(reader.uint32());
                        }
                    }
                    else {
                        message.clobPairId.push(reader.uint32());
                    }
                    break;
                case 2:
                    message.subaccountIds.push(subaccount_1.SubaccountId.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseStreamOrderbookUpdatesRequest();
        message.clobPairId = ((_a = object.clobPairId) === null || _a === void 0 ? void 0 : _a.map(e => e)) || [];
        message.subaccountIds = ((_b = object.subaccountIds) === null || _b === void 0 ? void 0 : _b.map(e => subaccount_1.SubaccountId.fromPartial(e))) || [];
        return message;
    }
};
function createBaseStreamOrderbookUpdatesResponse() {
    return {
        updates: []
    };
}
exports.StreamOrderbookUpdatesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.updates) {
            exports.StreamUpdate.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamOrderbookUpdatesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.updates.push(exports.StreamUpdate.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseStreamOrderbookUpdatesResponse();
        message.updates = ((_a = object.updates) === null || _a === void 0 ? void 0 : _a.map(e => exports.StreamUpdate.fromPartial(e))) || [];
        return message;
    }
};
function createBaseStreamUpdate() {
    return {
        blockHeight: 0,
        execMode: 0,
        orderbookUpdate: undefined,
        orderFill: undefined,
        takerOrder: undefined,
        subaccountUpdate: undefined
    };
}
exports.StreamUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.blockHeight !== 0) {
            writer.uint32(8).uint32(message.blockHeight);
        }
        if (message.execMode !== 0) {
            writer.uint32(16).uint32(message.execMode);
        }
        if (message.orderbookUpdate !== undefined) {
            exports.StreamOrderbookUpdate.encode(message.orderbookUpdate, writer.uint32(26).fork()).ldelim();
        }
        if (message.orderFill !== undefined) {
            exports.StreamOrderbookFill.encode(message.orderFill, writer.uint32(34).fork()).ldelim();
        }
        if (message.takerOrder !== undefined) {
            exports.StreamTakerOrder.encode(message.takerOrder, writer.uint32(42).fork()).ldelim();
        }
        if (message.subaccountUpdate !== undefined) {
            streaming_1.StreamSubaccountUpdate.encode(message.subaccountUpdate, writer.uint32(50).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.blockHeight = reader.uint32();
                    break;
                case 2:
                    message.execMode = reader.uint32();
                    break;
                case 3:
                    message.orderbookUpdate = exports.StreamOrderbookUpdate.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.orderFill = exports.StreamOrderbookFill.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.takerOrder = exports.StreamTakerOrder.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.subaccountUpdate = streaming_1.StreamSubaccountUpdate.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseStreamUpdate();
        message.blockHeight = (_a = object.blockHeight) !== null && _a !== void 0 ? _a : 0;
        message.execMode = (_b = object.execMode) !== null && _b !== void 0 ? _b : 0;
        message.orderbookUpdate = object.orderbookUpdate !== undefined && object.orderbookUpdate !== null ? exports.StreamOrderbookUpdate.fromPartial(object.orderbookUpdate) : undefined;
        message.orderFill = object.orderFill !== undefined && object.orderFill !== null ? exports.StreamOrderbookFill.fromPartial(object.orderFill) : undefined;
        message.takerOrder = object.takerOrder !== undefined && object.takerOrder !== null ? exports.StreamTakerOrder.fromPartial(object.takerOrder) : undefined;
        message.subaccountUpdate = object.subaccountUpdate !== undefined && object.subaccountUpdate !== null ? streaming_1.StreamSubaccountUpdate.fromPartial(object.subaccountUpdate) : undefined;
        return message;
    }
};
function createBaseStreamOrderbookUpdate() {
    return {
        snapshot: false,
        updates: []
    };
}
exports.StreamOrderbookUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.snapshot === true) {
            writer.uint32(8).bool(message.snapshot);
        }
        for (const v of message.updates) {
            off_chain_updates_1.OffChainUpdateV1.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamOrderbookUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.snapshot = reader.bool();
                    break;
                case 2:
                    message.updates.push(off_chain_updates_1.OffChainUpdateV1.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseStreamOrderbookUpdate();
        message.snapshot = (_a = object.snapshot) !== null && _a !== void 0 ? _a : false;
        message.updates = ((_b = object.updates) === null || _b === void 0 ? void 0 : _b.map(e => off_chain_updates_1.OffChainUpdateV1.fromPartial(e))) || [];
        return message;
    }
};
function createBaseStreamOrderbookFill() {
    return {
        clobMatch: undefined,
        orders: [],
        fillAmounts: []
    };
}
exports.StreamOrderbookFill = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.clobMatch !== undefined) {
            matches_1.ClobMatch.encode(message.clobMatch, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.orders) {
            order_1.Order.encode(v, writer.uint32(18).fork()).ldelim();
        }
        writer.uint32(26).fork();
        for (const v of message.fillAmounts) {
            writer.uint64(v);
        }
        writer.ldelim();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamOrderbookFill();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.clobMatch = matches_1.ClobMatch.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.orders.push(order_1.Order.decode(reader, reader.uint32()));
                    break;
                case 3:
                    if ((tag & 7) === 2) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.fillAmounts.push(reader.uint64());
                        }
                    }
                    else {
                        message.fillAmounts.push(reader.uint64());
                    }
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseStreamOrderbookFill();
        message.clobMatch = object.clobMatch !== undefined && object.clobMatch !== null ? matches_1.ClobMatch.fromPartial(object.clobMatch) : undefined;
        message.orders = ((_a = object.orders) === null || _a === void 0 ? void 0 : _a.map(e => order_1.Order.fromPartial(e))) || [];
        message.fillAmounts = ((_b = object.fillAmounts) === null || _b === void 0 ? void 0 : _b.map(e => helpers_1.Long.fromValue(e))) || [];
        return message;
    }
};
function createBaseStreamTakerOrder() {
    return {
        order: undefined,
        liquidationOrder: undefined,
        takerOrderStatus: undefined
    };
}
exports.StreamTakerOrder = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.order !== undefined) {
            order_1.Order.encode(message.order, writer.uint32(10).fork()).ldelim();
        }
        if (message.liquidationOrder !== undefined) {
            order_1.StreamLiquidationOrder.encode(message.liquidationOrder, writer.uint32(18).fork()).ldelim();
        }
        if (message.takerOrderStatus !== undefined) {
            exports.StreamTakerOrderStatus.encode(message.takerOrderStatus, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamTakerOrder();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.order = order_1.Order.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.liquidationOrder = order_1.StreamLiquidationOrder.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.takerOrderStatus = exports.StreamTakerOrderStatus.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseStreamTakerOrder();
        message.order = object.order !== undefined && object.order !== null ? order_1.Order.fromPartial(object.order) : undefined;
        message.liquidationOrder = object.liquidationOrder !== undefined && object.liquidationOrder !== null ? order_1.StreamLiquidationOrder.fromPartial(object.liquidationOrder) : undefined;
        message.takerOrderStatus = object.takerOrderStatus !== undefined && object.takerOrderStatus !== null ? exports.StreamTakerOrderStatus.fromPartial(object.takerOrderStatus) : undefined;
        return message;
    }
};
function createBaseStreamTakerOrderStatus() {
    return {
        orderStatus: 0,
        remainingQuantums: helpers_1.Long.UZERO,
        optimisticallyFilledQuantums: helpers_1.Long.UZERO
    };
}
exports.StreamTakerOrderStatus = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.orderStatus !== 0) {
            writer.uint32(8).uint32(message.orderStatus);
        }
        if (!message.remainingQuantums.isZero()) {
            writer.uint32(16).uint64(message.remainingQuantums);
        }
        if (!message.optimisticallyFilledQuantums.isZero()) {
            writer.uint32(24).uint64(message.optimisticallyFilledQuantums);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStreamTakerOrderStatus();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.orderStatus = reader.uint32();
                    break;
                case 2:
                    message.remainingQuantums = reader.uint64();
                    break;
                case 3:
                    message.optimisticallyFilledQuantums = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseStreamTakerOrderStatus();
        message.orderStatus = (_a = object.orderStatus) !== null && _a !== void 0 ? _a : 0;
        message.remainingQuantums = object.remainingQuantums !== undefined && object.remainingQuantums !== null ? helpers_1.Long.fromValue(object.remainingQuantums) : helpers_1.Long.UZERO;
        message.optimisticallyFilledQuantums = object.optimisticallyFilledQuantums !== undefined && object.optimisticallyFilledQuantums !== null ? helpers_1.Long.fromValue(object.optimisticallyFilledQuantums) : helpers_1.Long.UZERO;
        return message;
    }
};
